### PR TITLE
Support for connection to llamacpp server

### PR DESCRIPTION
**Pull request by JoseConseco**
This PR will allow to connect to [llamacpp server](https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md#unix-based-systems-linux-macos-etc), and fixes https://github.com/gsuuon/llm.nvim/issues/13 . This way  llama does not have to be started for each user prompt. This includes completions patch from issue mentioned above, provided by gsuuon. 
I'm new to this so fell free to fix any issues in this PR.

### CODE REVIEW

File: lua/llm/providers/llamacpp_server.lua [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  return curl.stream({
+    -- url = 'https://api-inference.huggingface.co/models/', --.. model,
+    url = "http://127.0.0.1:8080/completion",
+    method = "POST",
+    body = vim.tbl_extend("force", { stream = true }, params),
+    headers = {
+      -- Authorization = 'Bearer ' .. util.env_memo('HUGGINGFACE_API_KEY'),
+      ["Content-Type"] = "application/json",
+      -- ['data'] = '{"prompt": "Building a website can be done in 10 simple steps:","n_predict": 128}',
+    },
```
Comment (gsuuon) [inline lua/llm/providers/llamacpp_server.lua:26]: Headers aren't necessary here
```suggestion
```

File: lua/llm/providers/llamacpp_server.lua [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+local util = require "llm.util"
+local provider_util = require "llm.providers.util"
+
+local M = {}
+
+---@param handlers StreamHandlers
+---@param params? any Additional params for request
+---@param options? { model?: string }
+function M.request_completion(handlers, params, options)
+  local model = (options or {}).model or "bigscience/bloom"
```
Comment (gsuuon) [inline lua/llm/providers/llamacpp_server.lua:13]: We can get rid of all the huggingface stuff here
```suggestion
```

File: lua/llm/providers/llamacpp_server.lua [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  -- vim.print(params)
+
+  -- TODO handle non-streaming calls
+  return curl.stream({
+    -- url = 'https://api-inference.huggingface.co/models/', --.. model,
```
Comment (gsuuon) [inline lua/llm/providers/llamacpp_server.lua:18]:
```suggestion
```

File: lua/llm/providers/llamacpp_server.lua [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    provider_util.iter_sse_items(raw, function(item)
+      local data = util.json.decode(item)
+
+      if data == nil then
+        handlers.on_error(item, "json parse error")
+        return
+      end
+
+      if data.generation_settings ~= nil then -- last message
+        handlers.on_finish('', "stop")
```
Comment (gsuuon) [inline lua/llm/providers/llamacpp_server.lua:38]: Can just be a naked call now that it's handled in `provider.lua`
```suggestion
        handlers.on_finish()
```

File: lua/llm/providers/llamacpp_server.lua [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+
+M.default_prompt = {
+  provider = M,
+  options = {
+    -- model = 'bigscience/bloom'
+  },
+  params = {
+    return_full_text = false,
+  },
```
Comment (gsuuon) [inline lua/llm/providers/llamacpp_server.lua:103]: These are huggingface options / params, not needed here
```suggestion
```

**Review by gsuuon — COMMENTED** [review]
Almost there, just need to clean up some of the stuff pulled from the huggingface example. We should replace the llamacpp provider with this (so accept the changes and rename llamacpp_server.lua to llamacpp.lua)

Do you want to add some basic setup steps to the README for this? No worries if not!

File: lua/llm/providers/llamacpp_server.lua (hunk 3 above)
Comment (JoseConseco) [reply lua/llm/providers/llamacpp_server.lua]: done

File: lua/llm/providers/llamacpp_server.lua (hunk 5 above)
Comment (JoseConseco) [reply lua/llm/providers/llamacpp_server.lua]: done

File: lua/llm/providers/llamacpp_server.lua (hunk 4 above)
Comment (JoseConseco) [reply lua/llm/providers/llamacpp_server.lua]: done

**Comment by JoseConseco** [comment]
Demo of how this works currently (using my custom `modify` prompt): 
https://github.com/gsuuon/llm.nvim/assets/13521338/13802b45-2f66-4087-85e1-9c410f077559

**Comment by gsuuon** [comment]
Ok, lets get this merged. I'll replace the llamacpp provider and clean up. Thanks for contributing!

**Pull request merged**